¡Hola, Fercho!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un emnfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">


# Proyecto Final SQL

## 1. Descripción del objetivo del estudio.

Este proyecto se basa en un análisis de mercado de uno de los servicios de lectura y venta de libros que se lanzó durante la época de la pandemia. Contiene datos sobre los libros, editoriales y autores, entre otras cosas. Toda esta información se utilizará para generar una propuesta para el lanzamiento de un nuevo producto y así podremos saber si este será un éxito o un fracaso.

## 2. Conexión a la base de datos.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [6]:
db_config = {
    'user': 'practicum_student', # username
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # password
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com', 
    'port': 5432, # connection port
    'db': 'data-analyst-final-project-db' # the name of the database
}

In [7]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                        db_config['pwd'],
                                                        db_config['host'],
                                                        db_config['port'],
                                                        db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## 3. Exploración inicial.

In [34]:
query = """
SELECT *
FROM books
LIMIT 5;
"""

books_sample = pd.io.sql.read_sql(query, con = engine)
books_sample

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [10]:
query = """
SELECT *
FROM authors
LIMIT 5;
"""

authors_sample = pd.io.sql.read_sql(query, con = engine)
authors_sample

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


In [11]:
query = """
SELECT *
FROM publishers
LIMIT 5;
"""

publishers_sample = pd.io.sql.read_sql(query, con = engine)
publishers_sample

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


In [12]:
query = """
SELECT *
FROM ratings
LIMIT 5;
"""

ratings_sample = pd.io.sql.read_sql(query, con = engine)
ratings_sample

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


In [13]:
query = """
SELECT *
FROM reviews
LIMIT 5;
"""

reviews_sample = pd.io.sql.read_sql(query, con = engine)
reviews_sample

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


## 4. Análisis de mercado.

### Número de libros publicados después del 1 de enero del 2000.

In [17]:
query = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
"""
result = pd.io.sql.read_sql(query, con = engine)
print('Número de libros:')
print(result)

Número de libros:
   count
0    819


Después de verificar cuántos libros hay después del 2001, podemos ver que mas del 80% de estos fueron publicados despues de esta fecha.

### Número de reseñas de usuarios y la calificación promedio de cada libro.

In [22]:
query = """ 
SELECT title, 
       COUNT(review_id) AS total_reseñas, 
       AVG(rating) AS calificacion_promedio
FROM books 
JOIN reviews ON books.book_id = reviews.book_id
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY title, books.book_id
"""
result_cal = pd.io.sql.read_sql(query, con = engine)
print('Total de reseñas y sus promedios:\n')
print(result_cal)

Total de reseñas y sus promedios:

                                                 title  total_reseñas  \
0            The Body in the Library (Miss Marple  #3)              4   
1                                            Galápagos              4   
2                             A Tree Grows in Brooklyn             60   
3    Undaunted Courage: The Pioneering First Missio...              4   
4                                          The Prophet             28   
..                                                 ...            ...   
989                                Alice in Wonderland             52   
990          A Woman of Substance (Emma Harte Saga #1)              4   
991                                          Christine             21   
992  The Magicians' Guild (Black Magician Trilogy  #1)              4   
993                           The Plot Against America              4   

     calificacion_promedio  
0                 4.500000  
1                 4.500000  
2

Después de verificar cuántos libros tienen reseñas, podemos ver una gran participación en esto; prácticamente todos los libros tienen reseñas, aunque entre estos hay una gran varianza de entre 4 y 60 reseñas, y la calificación promedio de estos es bastante alta, lo cual también puede indicar una gran aceptación de los usuarios a los libros publicados.

### Nombres de editoriales que tienen más libros con más de 50 páginas.

In [35]:
query = """
SELECT publisher, COUNT(*)
FROM books 
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50 
GROUP BY publishers.publisher
ORDER BY COUNT(*) DESC
"""

result_50 = pd.io.sql.read_sql(query, con = engine)
print('Editoriales que tienen mayor numero de libros con mas de 50 paginas:\n')
print(result_50)

Editoriales que tienen mayor numero de libros con mas de 50 paginas:

                                             publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]


Después de verificar entre las editoriales (334) y aplicarles un filtro para que solo se calificaran a editoriales en que sus libros tengan más de 50 páginas, podemos ver que Penguin Books es la editorial con la mayor cantidad de estos, con 42 libros, seguido por Vintage con 31 y Grand Central Publishing con 25. 

## Autor que tiene la más alta calificación (al menos 50 calificaciones registradas).

In [26]:
query = """
SELECT author AS autor, AVG(rating) AS promedio
FROM books
JOIN authors ON books.author_id = authors.author_id
JOIN ratings ON books.book_id = ratings.book_id
GROUP BY author
HAVING COUNT(rating) > 50
"""

result_author = pd.io.sql.read_sql(query, con = engine)
print('Promedio de las calificaciones de los autores con 50 calificaciones:\n')
print(result_author)

Promedio de las calificaciones de los autores con 50 calificaciones:

                                                autor  promedio
0                          J.K. Rowling/Mary GrandPré  4.288462
1                                           Dan Brown  3.741259
2                                       J.D. Salinger  3.846939
3                                     Agatha Christie  4.283019
4             George Orwell/Boris Grabnar/Peter Škerl  3.729730
5                                      John Steinbeck  3.643836
6                                     William Golding  3.901408
7                                     Nicholas Sparks  3.882883
8                                       Arthur Golden  4.107143
9                                        John Grisham  3.971429
10                                       Stephen King  4.009434
11                                       Jodi Picoult  3.881579
12  William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
13                                

Podemos ver que solo 22 autores cuentan con más de 50 calificaciones en sus libros y estos cuentan con calificaciones más altas de la media. Destacando, J.K. Rowling tiene la calificación más alta (4.29). Agatha Christie también está muy bien valorada (4.28). J. R. R. Tolkien tiene una excelente calificación (4.24), pero notemos que John Steinbeck tiene la calificación más baja del grupo (3.64). Esto nos da una idea de la aceptación de los usuarios, la cual podemos concluir que es buena.

## Número promedio de reseñas entre los usuarios que calificaron mas de 50 libros.

In [32]:
query = """
SELECT reviews.username, 
    COUNT(reviews.book_id) AS numere_de_calificaciones, 
    AVG(rating) AS promedio
FROM ratings
JOIN reviews ON ratings.book_id = reviews.book_id 
GROUP BY reviews.username
HAVING COUNT(reviews.book_id) > 50
"""
result_user = pd.io.sql.read_sql(query, con = engine)
print('Calificacion promedio de usuarios que hicieron mas de 50 reseñas:\n')
print(result_user)

Calificacion promedio de usuarios que hicieron mas de 50 reseñas:

           username  numere_de_calificaciones  promedio
0         zjohnston                       199  4.160804
1           jnelson                       302  3.910596
2      lewisdesiree                       172  4.058140
3     patrickhudson                       141  3.680851
4      valenciaanne                       293  4.003413
..              ...                       ...       ...
151  shermannatalie                       144  3.756944
152         dmiller                       104  3.884615
153  charlesraymond                        78  3.923077
154     isaiahreyes                       152  4.203947
155       masonsara                        79  3.873418

[156 rows x 3 columns]


Podemos ver que los usuarios dejan bastantes reseñas en los libros que leen y el promedio de las calificaciones es bastante alto para la mayoría de los libros, aunque podemos decir que no hay un patrón a seguir en esta consulta, ya que hay usuarios con muchas reseñas y baja calificación promedio, y lo mismo pasa por el contrario, pero podemos decir que los usuarios aceptan positivamente los libros que se publican en la plataforma.

<div class="alert alert-block alert-success">
<b>Paso 2</b> <a class="tocSkip"></a><br>
Has realizado un buen trabajo al formular una consulta SQL para contar con:
    
    - el número de libros publicados después de una fecha específica.
    - Calificación promedio
    - Editorial que ha publicado la mayor cantidad de libros
    - Autor con la calificación promedio más alta
    - Promedio de reseñas
    
Esto demuestra tu habilidad para manipular y consultar bases de datos de manera efectiva. Solamente para terminar de complementar te recomiendo incluir el número de libros evaluados por autor para proporcionar un contexto adicional sobre la consistencia de sus calificaciones.
</div>